First batch of Sematic annotations (Alpha2)

dennwc opened the pull request:
Implements all low-hanging fruits in Semantic annotations:
- `Identifier`
- `String`
- `Block`
- `Import`
- `QualifiedIdentifier`

It also implements the positional information preprocessor that is common for Semantic and Annotated modes.

This leaves only `Comment` (see #3) and `Function` in TODO list for Semantic mode.

PR takes few shortcuts by dropping some additional positional info that C# parser provides (like bracket postions). Those can be included to the UAST later.

creachadair commented on driver/normalizer/normalizer.go:
```diff
@@ -23,14 +16,178 @@ var Normalize = Transformers([][]Transformer{
-	// (like node type, token, positional information).
-	//
-	// https://godoc.org/gopkg.in/bblfsh/sdk.v2/uast#ObjectToNode
+	// replace "IsEmpty" with "Length: 0" in spans
```
Throughout, could you please add brief one-liners to each of the transforms to say what it's generally intended to do? I realize some of these may be fairly standard, but it's hard to decipher the intent from the literal syntax. (This first one says what it's doing—but not why)

creachadair commented on fixtures/Program.cs:
```diff
@@ -0,0 +1,106 @@
+using System;
+using System.Linq;
```
Is this used?

dennwc replied:
Yes, it adds `Where` extension to `IList`.

creachadair replied:
Ugh. Implicit side-effects are the worst.  :)

creachadair commented on fixtures/Program.cs:
```diff
@@ -0,0 +1,106 @@
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+using Microsoft.CodeAnalysis;
```
Is this used?

dennwc replied:
Yes, some interface types are from this package.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -23,14 +16,178 @@ var Normalize = Transformers([][]Transformer{
+		}),
+	),
+
+	// restore "Start" if it was removed as a default value
```
Well, I enabled a special mode in JSON encoder to skip fields that are equal to their default values. The upside is that the native AST is significantly smaller this way. The downside is that it becomes a bit harder to work with if it cuts a field required by the DSL statements.

creachadair replied:
Ah, I understand. It makes sense. 
Could you please note that in a comment here?

juanjux replied:
Mmmm I did something like that with the C++ driver and we finally decided that it was better for the user to not drop those fields. Probably applies here too.

bzz commented on driver/normalizer/normalizer.go:
```diff
@@ -23,14 +16,178 @@ var Normalize = Transformers([][]Transformer{
+				"Value":          Var("name"),
+				"ValueText":      Var("name"),
+			},
+			"RawKind": Int(8616),
```
May be that is something obvious, but what are the `RawKind` constants for?

dennwc replied:
This is probably an enum for a parser's rule id. I will check if it's possible to change the JSON encoder to get the textual representation for it.

Another question you may ask is why it's mentioned here at all. The reason is the following: the DSL is very strict to fields that appear in AST, but was not covered by the transformation rule. This was done to avoid cases when we silently drop unknown fields from AST nodes. So each rule should list every single field in the AST node either as a variable or as a constant. Unless the transform is `Part`ial, as in most preprocessor transforms.

juanjux commented:
Looks very good, congratulations!

I think it could be slimmed even more. I see a lot of empty but noisy `SyntaxTrivia` nodes that doesn't have anything except the position. Also the `CloseBraceToken`s and `SemiColonToken`s could also be removed (curious note: most of these have an ending position on a different line).

dennwc commented:
Re-enabled the default values and added more comments to annotation code as requested.

@juanjux Right, the `Trivia` nodes are mostly whitespaces, but it seems like it may be related to comments as well. I will try to cut as much of those nodes as I can, but I don't want to disable all of them.

dennwc commented:
Also note that it's blocked by https://github.com/bblfsh/sdk/pull/345.

juanjux commented:
Another thing that could be done in this or other PR is removing the "Syntax" postfix from nodes types since almost all have it.

juanjux commented:
Things that have been showing while adding tests: definitely the RawKind must be translated into a string. It's using this enum:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntaxkind?view=roslyn-dotnet

The reason to be translated, is that it specifies types not in the current @type. For example, on the arithmetic operators test, only the `RawKinds` tell the plus operator or the `AddExpression`.

dennwc commented:
Updated all the fixtures (thanks again, @tsolakoua!) and reverted an update to the unreleased SDK. New functions are now unnecessary because all the default fields are present and there is no need to write DSL conditions for them.

@juanjux Thanks for the link, I will definitely convert them to strings one way or another. Also the proposal to drop "Syntax" sounds reasonable. But I'd like to merge this PR and continue in the next one, if there are no objections :)
